src/App/Fossa/VSI/Fingerprint.hs

```diff
@@ -0,0 +1,217 @@
+  fingerprint,
+  Fingerprint,
+  Raw,
+  CommentStripped,
+  Combined (..),
+) where
+
+import Conduit (ConduitT, Void, await, decodeUtf8C, encodeUtf8C, filterC, linesUnboundedC, mapC, runConduitRes, sourceFile, yield, (.|))
+import Control.Algebra (Has)
+import Control.Effect.Diagnostics (Diagnostics, fatalText)
+import Control.Effect.Exception (IOException, Lift, catch)
+import Control.Effect.Lift (sendIO)
+import Crypto.Hash (Digest, HashAlgorithm, SHA256 (..), hashFinalize, hashInit, hashUpdate)
+import Data.Aeson (ToJSON, object, toJSON, (.=))
+import Data.ByteString qualified as B
+import Data.Maybe (fromMaybe)
+import Data.String.Conversion (ToText (..))
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Text.Extra (breakOnAndRemove)
+import Discovery.Walk (WalkStep (..), walk')
+import Effect.ReadFS (ReadFS, contentIsBinary)
+import Path (Abs, Dir, File, Path, toFilePath)
+
+-- | Fingerprint deterministically idenfies a file and is derived from its content.
+--
+-- The type variable is:
+--
+--   @k@ - Kind, the kind of fingerprint computed.
+--
+-- For ease of implementation, the backing representation of a @Fingerprint@ instance is a @Base16@ encoded @Text@.
+newtype Fingerprint k = Fingerprint Text
+
+type role Fingerprint nominal
+
+-- | Represents a 'Fingerprint' derived from the unmodified content of a file.
+data Raw
+
+-- | Represents a 'Fingerprint' derived from the content of a file with all C-style comments removed.
+data CommentStripped
+
+instance ToText (Fingerprint k) where
+  toText (Fingerprint k) = k
+
+instance ToJSON (Fingerprint k) where
+  toJSON = toJSON . toText
+
+-- | Represents the result of running all fingerprinting implementations on a file.
+data Combined = Combined
+  { combinedRaw :: Fingerprint Raw
+  , combinedCommentStripped :: Maybe (Fingerprint CommentStripped)
+  }
+
+instance ToJSON Combined where
+  toJSON Combined{..} =
+    object
+      [ "sha_256" .= toText combinedRaw
+      , "comment_stripped:sha_256" .= fmap toText combinedCommentStripped
+      ]
+
+encodeFingerprint :: Digest hash -> Fingerprint t
+encodeFingerprint = Fingerprint . toText . show
+
+-- | Hashes a stream of 'B.ByteString'@s@ and creates a digest @d@.
+-- Adapted from @sinkHash@ in https://hackage.haskell.org/package/cryptonite-conduit-0.2.2/docs/src/Crypto-Hash-Conduit.html
+sinkHash :: (Monad m, HashAlgorithm hash) => ConduitT B.ByteString Void m (Digest hash)
+sinkHash = sink hashInit
+  where
+    sink ctx = do
+      b <- await
+      case b of
+        Nothing -> return $! hashFinalize ctx
+        Just bs -> sink $! hashUpdate ctx bs
+
+-- | Hashes the whole contents of the given file in constant memory.
+hashBinaryFile :: (Has (Lift IO) sig m, Has Diagnostics sig m, HashAlgorithm hash) => FilePath -> m (Digest hash)
+hashBinaryFile fp =
+  sendIO (runConduitRes (sourceFile fp .| sinkHash))
+    `catch` (\(e :: IOException) -> fatalText ("unable to hash file: " <> toText (show e)))
```

Comment:
Can we extract this to something like `catchingIOExc :: m a -> m a`?  It's used multiple times.  This might even belong in `Control.Effect.Diagnostics` (with a parameterized type and generic message).